ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -58,6 +58,30 @@ let ensure_domain_type_is_specified ~__context ~self =
     derive_domain_type ~hVM_boot_policy:hbp |> fun value ->
     Db.VM.set_domain_type ~__context ~self ~value
 
+let ensure_boot_params_consistent ~__context ~self =
+  (* XSI-804 if we are missing an HVM boot order, then qemu won't
+   * be able to start the VM successfully *)
+  if Db.VM.get_domain_type ~__context ~self = `hvm then
+    let boot_params = Db.VM.get_HVM_boot_params ~__context ~self in
+    let msg =
+      match List.assoc_opt "order" boot_params with
+      | None ->
+          Some "is missing HVM_boot_params.order"
+      | Some "" ->
+          Some "has empty HVM_boot_params.order"
+      | Some _ ->
+          None
+    in
+    match msg with
+    | None ->
+        ()
+    | Some msg ->
+        raise
+          Api_errors.(
+            Server_error
+              ( internal_error
```

Comment:
If this is an internal error, is there something the user can do to fix the root cause?

Can xapi try to recover from this and give a default for boot params' order?

Reply:
Xapi could guess the boot order, but I'm not sure that's a good idea

I was expecting that the user would call something like `xe vm-param-set HVM-boot-params:order=cdn`

Reply:
It might be worth creating an API error for this situation so xencenter can have more control over this.

@kc284 what do you think?

Reply:
We have https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_xenops.ml#L404, which fills in `"cd"` by default, but only if `order` is missing from the map rather than empty. We could change that to put in the default for `""` as well.

Reply:
If we raise an error, then it should not be an internal error, because that implies a bug rather than a problem with user input. You could use `invalid_value`.

Reply:
Ah I didn't see your comment earlier about xapi_xenops filling in the order parameter with a default. That sounds like a better solution to me - as long as assuming a boot order isn't going to make the VM explode

Reply:
Can we add this default value to `ocaml/xapi-consts/constants.ml` and use it in all sites?

```
ocaml/xapi-consts/constants.ml
164:let hvm_boot_params_order = "order"

xen-api/ocaml/xapi/import_xva.ml
74:      let hVM_boot_params = if vm.is_hvm then [("order", "cd")] else [] in
xen-api/ocaml/xapi/xapi_xenops.ml
404:    ; boot_order= string vm.API.vM_HVM_boot_params "cd" "order"
```

Reply:
I will, though this import xva code is going to be removed soon iirc